src/backend/distributed/utils/metadata_cache.c

```diff
@@ -2598,19 +2598,28 @@ InitializeCaches(void)
 
 	if (!performedInitialization)
 	{
-		/* set first, to avoid recursion dangers */
-		performedInitialization = true;
-
 		/* make sure we've initialized CacheMemoryContext */
 		if (CacheMemoryContext == NULL)
 		{
 			CreateCacheMemoryContext();
 		}
 
 		InitializeDistTableCache();
+
+		/*
+		 * Make sure DistShardCacheHash is created successfully before setting
+		 * performedInitialization flag and registering callbacks.
+		 */
+		if (DistShardCacheHash != NULL)
```

Comment:
I don't think this can really happen. What could happen is that we throw an OOM before we get to this point that leaves the cache in a partially initialised state. Moving the boolean assignment should cause us to reinitialise.

What we might also consider doing is cleaning up the memory when the cache is partially initialised before re-initialising it, e.g. we could use a separate memory context (child of CacheMemoryContext) and reset that in a PG_CATCH. That might be overkill since this should be a very rare event.

Reply:
One way partial initialization can happen is DistTableCacheHash could be created successfully [here](https://github.com/citusdata/citus/blob/master/src/backend/distributed/utils/metadata_cache.c#L2646) but following [call](https://github.com/citusdata/citus/blob/master/src/backend/distributed/utils/metadata_cache.c#L2653) could fail.

I still think DistShardCacheHash remains unset here.

Reply:
If the call fails I would expect an error to be thrown, hence we don't continue with this function.

Reply:
done